pkg/http/bufconn.go

```diff
@@ -0,0 +1,88 @@
+/*
+ * Minio Cloud Storage, (C) 2017 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package http
+
+import (
+	"bufio"
+	"net"
+	"time"
+)
+
+// bufConn - is a generic stream-oriented network connection supporting buffered reader and read/write timeout.
+type bufConn struct {
+	net.Conn
+	bufReader              *bufio.Reader // buffered reader wraps reader in net.Conn.
+	readTimeout            time.Duration // sets the read timeout in the connection.
+	writeTimeout           time.Duration // sets the write timeout in the connection.
+	updateBytesReadFunc    func(int)     // function to be called to update bytes read.
+	updateBytesWrittenFunc func(int)     // function to be called to update bytes written.
+}
+
+func (c *bufConn) setReadTimeout() {
+	if c.readTimeout != 0 {
```

Comment:
You should stop setting readTimeouts after the fact that we have started reading from the Conn successfully.  The reason being that it doesn't really help to have deadlines in this manner which provide no proper control. 

This is carefully visible in my previous comment on how net/rpc can disconnect and requires continuous reconnects every 30secs.